Snap CupertinoSliverNavigationBar large title and bottom widget

victorsanni opened the pull request:
Fixes [CupertinoSliverNavigationBar should snap between large and middle title](https://github.com/flutter/flutter/issues/126028)
Fixes [CupertinoSliverNavigationBar needs to implement snapping](https://github.com/flutter/flutter/issues/23321)

## Flutter (with this PR):
https://github.com/user-attachments/assets/b0177c06-e4b5-4981-8ec3-62171a4f2547

## Native:
https://github.com/user-attachments/assets/6efa0825-d26c-4c20-81cc-c9fada6e15fc

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

victorsanni committed:
Snap CupertinoSliverNavigationBar large title

victorsanni committed:
Fix crash on desktop

victorsanni committed:
Merge branch 'master' into snap-large-title

victorsanni committed:
Add tests

victorsanni committed:
Merge branch 'master' into snap-large-title

victorsanni committed:
Tweak scroll gesture name

victorsanni committed:
Tweak test comments

victorsanni committed:
Merge branch 'master' into snap-large-title

MitchellGoodwin reviewed (commented):
Was ScrollEndNotification unable to be used for this? Just following up on this comment: https://github.com/flutter/flutter/issues/126028#issuecomment-2305569145

victorsanni commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -818,13 +818,44 @@ class CupertinoSliverNavigationBar extends StatefulWidget {
+          : 0.0,
+        // Eyeballed on an iPhone 16 simulator running iOS 18.
+        duration: const Duration(milliseconds: 100),
+        curve: Curves.easeOut,
```
Yeah, tbh I don't know what animation curve or values are used. The native one seems more snappy, but not too fast at the same time. If only there was a way to get these values...

Piinks replied:
Maybe `Curves.easeOutExpo`? That one seems to start off faster. There is a helpful video with graphs and sample animation for every curve at https://api.flutter.dev/flutter/animation/Curves-class.html

MitchellGoodwin replied:
I have a strong suspicion that it's a spring animation. That's what iOS uses for releasing a partially dragged page transition. There is a way to check with Xcode debugging, but I do not remember off the top of my head. @dkwingsmt was working with using the spring simulation for an animation: https://github.com/flutter/flutter/pull/155575

But my opinion for this PR is it's fine to use a curve we already have for now, but one that's closer. [fastEaseInToSlowEaseOut](https://api.flutter.dev/flutter/animation/Curves/fastEaseInToSlowEaseOut-constant.html) is what the Cupertino routes use.

victorsanni replied:
easeOutExpo started a little bit too fast, but fastEaseInToSlowEaseOut was closer-ish. It's not perfect but definitely better.

victorsanni commented:
> Was ScrollEndNotification unable to be used for this? 

I was having difficulty making scroll notifications and scroll position play well together. I added a listener to be called on `ScrollNotification`s, but when I tried to animate the scroll position within that listener, it didn't work. So I added a separate listener for the scroll position, and tried making both callbacks talk to each other, but that also didn't work.

I agree that responding to `ScrollEndNotification`s seems like the most intuitive way to do this. I think Hans' examples worked with `ScrollNotification`s because it had control of the parent `CustomScrollView` and so could wrap it directly in a `NotificationListener`. But it's trickier to do (but maybe there is a way to do it) when we're entirely within a sliver within the `CustomScrollView` and can only look up the tree with inherited widgets.

Piinks commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -818,13 +818,44 @@ class CupertinoSliverNavigationBar extends StatefulWidget {
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _scrollableState?.position.isScrollingNotifier.removeListener(_handleScrollChange);
```
Using the `isScrollingNotifier` makes sense to me. It will only notify when scrolling starts and stops, which is exactly what is needed here. 👍  Using scroll notifications would mean a lot of notifications we would not need.

Piinks commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -818,13 +818,44 @@ class CupertinoSliverNavigationBar extends StatefulWidget {
+
+  void _handleScrollChange() {
+    final ScrollPosition? position = _scrollableState?.position;
+    if (position != null && position.pixels > 0.0 && position.pixels < _kNavBarLargeTitleHeightExtension) {
```
You might want to add `position.hasPixels` before accessing `position.pixels`, otherwise it will throw.

victorsanni committed:
Add position.hasPixels

victorsanni committed:
Merge branch 'master' into snap-large-title

victorsanni committed:
Tweak animation values

victorsanni committed:
Merge branch 'master' into snap-large-title

victorsanni committed:
Remove trailing spaces

MitchellGoodwin reviewed (approved):
LGTM!

victorsanni commented:
Converting to draft... Since the bottom to nav bar PR has been merged, I'm working to make it snap for both the search bar and the large title.

victorsanni commented:
Snapping only the large title causes problems in `NavigationBarBottomMode.automatic`, since the search bar scrolls first instead of the large title (so we should be using the search bar height to snap first, not the large title height).

victorsanni committed:
Merge branch 'master' into snap-large-title

victorsanni committed:
Add functionality and tests for bottom widget

victorsanni committed:
Merge branch 'master' into snap-large-title

victorsanni committed:
Clip bottom in automatic mode

victorsanni committed:
Merge branch 'master' into snap-large-title

MitchellGoodwin commented on packages/flutter/test/cupertino/nav_bar_test.dart:
```diff
@@ -2087,6 +2087,212 @@ void main() {
+    await tester.pumpAndSettle();
+
+    // Expect the bottom to snap up to the large title.
+    expect(scrollController.position.pixels, bottomHeight);
```
Should probably add a check that the bottom widget is hidden directly as well, for completion sake. Instead of just the scroll position.

MitchellGoodwin commented on packages/flutter/test/cupertino/nav_bar_test.dart:
```diff
@@ -2087,6 +2087,212 @@ void main() {
+
+    // Expect the large title to snap back to its extended height, which is the
+    // same scroll offset as the fully-shrunk bottom widget.
+    expect(scrollController.position.pixels, bottomHeight);
```
Same here. Should add a more direct check that the bottom widget is not visible.

MitchellGoodwin reviewed (approved):
Renewing my LGTM with some nits.

victorsanni committed:
Ensure correct bottom position in tests

victorsanni committed:
Merge branch 'master' into snap-large-title

victorsanni committed:
Remove unneeded MediaQuerys

victorsanni committed:
Merge branch 'master' into snap-large-title

Piinks reviewed (approved):
LGTM

b3nni97 commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -921,13 +921,61 @@ class CupertinoSliverNavigationBar extends StatefulWidget {
+    }
+
+    if (target != null) {
+      position.animateTo(
```
Using 'animateTo' can be very risky as it breaks the entire scrollable widget if one attempts to scroll within the 300ms timeframe (refer to issue #14452). This issue is easily reproducible by increasing the animation duration and trying to scroll the CustomScrollView during that period. Subsequently, assertions fail, such as '_hold == null || _drag == null'.
